libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,581 @@
+
+#define REGISTER_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, static_check_wl_args, context_aware_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_CF_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, static_check_wl_args, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, context_aware_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_CF_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+#define REGISTER_LEGACY_CF_ONLY_HOST_FUNCTION(NAME, ...) \
+   inline static host_function_registrator<&interface::NAME, alias_check, early_validate_pointers, legacy_static_check_wl_args, context_free_check, ##__VA_ARGS__> NAME ## _registrator = {"env", #NAME};
+
+   class interface {
+      public:
+         interface(apply_context& ctx) : context(ctx) {}
+
+         inline apply_context& get_context() { return context; }
+         inline const apply_context& get_context() const { return context; }
+
+         // context free api
+         int32_t get_context_free_data(uint32_t index, legacy_array_ptr<char> buffer) const;
+         REGISTER_LEGACY_CF_ONLY_HOST_FUNCTION(get_context_free_data)
+
+         // privileged api
+         int32_t is_feature_active(int64_t feature_name) const;
+         void activate_feature(int64_t feature_name) const;
+         void preactivate_feature(const digest_type&);
+         void set_resource_limits(account_name account, int64_t ram_bytes, int64_t net_weight, int64_t cpu_weight);
+         void get_resource_limits(account_name account, legacy_ptr<int64_t, 8> ram_bytes, legacy_ptr<int64_t, 8> net_weight, legacy_ptr<int64_t, 8> cpu_weight) const;
+         int64_t set_proposed_producers(legacy_array_ptr<char> packed_producer_schedule);
+         int64_t set_proposed_producers_ex(uint64_t packed_producer_format, legacy_array_ptr<char> packed_producer_schedule);
+         uint32_t get_blockchain_parameters_packed(legacy_array_ptr<char> packed_blockchain_parameters) const;
+         void set_blockchain_parameters_packed(legacy_array_ptr<char> packed_blockchain_parameters);
+         bool is_privileged(account_name account) const;
+         void set_privileged(account_name account, bool is_priv);
+
+         REGISTER_HOST_FUNCTION(is_feature_active, privileged_check);
+         REGISTER_HOST_FUNCTION(activate_feature, privileged_check);
+         REGISTER_HOST_FUNCTION(preactivate_feature, privileged_check);
```

Comment:
This should be legacy. We don't want to support `const digest_type&` as an argument for host functions going forward. 

Also, I'm concerned about the possible changed behavior of references. I think we may need a new `legacy_ref` wrapper type and make this host function's argument type into `legacy_ref<const digest_type>`.

Reply:
No need for `legacy_ref`. This argument would type would instead be `legacy_ptr<const digest_type>`.